cmd/distrogen/component.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+var errInvalidComponentType = errors.New("invalid component type")
+
+type ComponentGenerator struct {
+	Spec     *DistributionSpec
+	FileMode fs.FileMode
+
+	Type      ComponentType
+	Name      string
+	Path      string
+	ModuleURL string
+}
+
+func NewComponentGenerator(spec *DistributionSpec, componentType ComponentType, componentName string) (*ComponentGenerator, error) {
+	if spec.ComponentModuleBase == "" {
+		return nil, errors.New("must supply a component_module_base in spec")
+	}
+
+	g := &ComponentGenerator{
+		Spec:     spec,
+		FileMode: DefaultProjectFileMode,
+		Type:     componentType,
+		Name:     componentName,
+	}
+
+	switch componentType {
+	case Receiver:
+		fallthrough
+	case Processor:
+		fallthrough
+	case Exporter:
+		fallthrough
+	case Connector:
+		fallthrough
+	case Extension:
+		fallthrough
+	case Provider:
+		g.Path = filepath.Join(
+			"components",
+			string(componentType),
+			fmt.Sprintf("%s%s", componentName, componentType),
+		)
+
+	default:
+		return nil, fmt.Errorf("%w: %s", errInvalidComponentType, componentType)
+	}
+
+	g.ModuleURL = path.Join(spec.ComponentModuleBase, g.Path)
+
+	return g, nil
+}
+
+func (g *ComponentGenerator) Generate() error {
+	componentTemplates, err := GetIndividualComponentTemplateSet(g, g.FileMode)
+	if err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(g.Path, g.FileMode); err != nil {
+		logger.Debug("generator", "generator", *g)
```

Comment:
These debug logs can probably be removed when you cleanup and make the PR ready for full review.